src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -16,168 +17,383 @@
+
+		//! allocates a chunk from which STightStructs::COUNT blocks will be suballocated, required structs are indirects & elements
+		bool allocateRequiredBlock(mdi_buffer_t* mdi)
+		{
+			requiredStructsBlockInfo.allocated = true;
+
+			auto [blockOffset, blockSize] = std::make_tuple(InvalidAddress, std::min(mdi->compose->max_size(), std::reduce(std::begin(requiredStructsBlockInfo.sizes), std::end(requiredStructsBlockInfo.sizes))));
+
+			mdi->compose->multi_allocate(std::chrono::steady_clock::now() + std::chrono::microseconds(100u), 1u, &blockOffset, &blockSize, &MdiMaxAlignment);
+
+			if (blockOffset == InvalidAddress)
+				return (requiredStructsBlockInfo.allocated = false);
+
+			bigChunkRequestInfo.offsets.emplace_back() = blockOffset;
+			bigChunkRequestInfo.sizes.emplace_back() = blockSize;
+
+			auto* const mdiData = reinterpret_cast<uint8_t*>(mdi->compose->getBufferPointer());
+			mdi_buffer_t::suballocator_traits_t::allocator_type fillSubAllocator(mdiData, blockOffset, 0u, MdiMaxAlignment, blockSize);
+			mdi_buffer_t::suballocator_traits_t::multi_alloc_addr(fillSubAllocator, requiredStructsBlockInfo.offsets.size(), requiredStructsBlockInfo.offsets.data(), requiredStructsBlockInfo.sizes.data(), requiredStructsBlockInfo.alignments.data());
+
+			for (const auto& offset : requiredStructsBlockInfo.offsets)
+				if (offset == InvalidAddress)
+					return (requiredStructsBlockInfo.allocated) = false;
+
+			return requiredStructsBlockInfo.allocated;
+		}
+
+		void latchDeallocations(mdi_buffer_t* mdi, ISemaphore::SWaitInfo waitInfo)
+		{
+			mdi->compose->multi_deallocate(bigChunkRequestInfo.offsets.size(), bigChunkRequestInfo.offsets.data(), bigChunkRequestInfo.sizes.data(), waitInfo);
+		}
+
+		inline const auto& getLimits() { return limits; }
+		inline const auto& getRequiredStructsBlockInfo() { return requiredStructsBlockInfo; }
+
+		struct
+		{
+			std::vector<mdi_size_t> offsets, sizes;
+		} bigChunkRequestInfo;
+
+	private:
+
+		struct SLimits
+		{
+			//! sum of  metaList[x].sizes - all bytes which needs to be uploaded to cover all of totalIndirectDrawCount objects, note we don't count element & indirect structers there
+			size_t totalByteSizeRequest = {},
+
+			//! amount of total objects to draw with indirect indexed call
+			totalIndirectDrawCount = {};
+		};
+
+		const ImDrawData* drawData;
+		std::vector<DrawItemMeta> metaList;
+
+		SLimits limits;
+		STightStructs requiredStructsBlockInfo;
+	};
+
+	static constexpr SPushConstantRange PushConstantRanges[] =
+	{
+		{
+			.stageFlags = IShader::E_SHADER_STAGE::ESS_VERTEX | IShader::E_SHADER_STAGE::ESS_FRAGMENT,
+			.offset = 0,
+			.size = sizeof(PushConstants)
+		}
+	};
+
+	smart_refctd_ptr<IGPUPipelineLayout> UI::createDefaultPipelineLayout(IUtilities* const utilities, const SResourceParameters::SBindingInfo texturesInfo, const SResourceParameters::SBindingInfo samplersInfo, uint32_t texturesCount)
+	{
+		if (!utilities)
+			return nullptr;
```

Comment:
why do you need `utilities` to create a pipeline layout!?

Reply:
resolved